Sanitize HTML (#5580)
dotnet/docfx: docs/designs/sanitize-html.md

### Sanitization Modes

There are 3 sanitization modes:

- `standard` (*default*): Described below
- `strict`: Disallow html
- `loose`: This is based on the `standard` mode, with a few more allowed HTML attributes like `class`, `style`. The final list came from the telemetry of the old hub/landing pages and archived contents.

Sanitization mode is automatically set to `loose` for old hub pages, landing pages and archived contents.

Sanitization mode can be configured per JSON schema property:
```json
{
  "contentType": "markdown",
  "htmlSanitizationMode": "strict",
}
```

### Standard HTML tags and attributes sanitization

Use an _allow list_ to sanitize HTML tags and attributes. HTML tags can contain the follow common attributes:

- Attribute names starting with `data-`.
- Accessibility attributes: `role` and names starting with `aria-`.
- Standard HTML5 global attributes:

  ```html
  name, id, itemid, itemprop, itemref, itemscope, itemtype,
  part, slot, spellcheck, title
  ```

Each allowed HTML tag can have additional attributes as follows:

> > NOTE: This list may be adjustmented based on docs usage collected from telemetry

Reviewer: "adjustmented" is not an English word. Should be "adjusted"

Author: oops, my silly typo, Thanks!

### Sanitization Modes

There are 3 sanitization modes:

- `standard` (*default*): Described below
- `strict`: Disallow html
- `loose`: This is based on the `standard` mode, with a few more allowed HTML attributes like `class`, `style`. The final list came from the telemetry of the old hub/landing pages and archived contents.

Sanitization mode is automatically set to `loose` for old hub pages, landing pages and archived contents.

Sanitization mode can be configured per JSON schema property:
```json
{
  "contentType": "markdown",
  "htmlSanitizationMode": "strict",
}
```

### Standard HTML tags and attributes sanitization

Use an _allow list_ to sanitize HTML tags and attributes. HTML tags can contain the follow common attributes:

- Attribute names starting with `data-`.
- Accessibility attributes: `role` and names starting with `aria-`.
- Standard HTML5 global attributes:

  ```html
  id, itemid, itemprop, itemref, itemscope, itemtype,
  part, slot, spellcheck, title
  ```

Each allowed HTML tag can have additional attributes as follows:

```html
<!-- Content sectioning -->
<address>: 

> <h1>-<h6>: 

Author: There is a validation feature to block h1-h6 entirely for user written HTML, I'm removing them entirely from here.

<section>:

<!-- Text content -->
<blockquote>: cite
<dd>:
<div>:
<dl>:
<dt>:
<figcaption>:
<figure>:
<hr>:
<li>: value
<ol>: reversed, start, type
<p>:
<pre>:
<ul>:

<!-- Inline text semantics -->

> <a>: download, href, hreflang, ping, rel, target, type

Reviewer 2: `href` only in both loose and strict modes

<abbr>:
<b>:
<bdi>:
<bdo>:
<br>:
<cite>:
<code>:

> <data>: value

Reviewer 2: not needed?

<dfn>:
<em>:
<i>:
<mark>:
<q>: cite
<s>:
<samp>:
<small>:
<span>:
<strong>:
<sub>:
<sup>:
<time>: datetime
<u>:
<var>:

<!-- Image and multimedia -->

> <img>: alt, decoding, height, intrinsicsize, loading, sizes, src, width

Reviewer 2: `alt`, `src`, `width`, `height`

<!-- Demarcating edits -->
<del>: cite, datetime
<ins>: cite, datetime

<!-- table -->
<caption>:
<col>:
<colgroup>:
<table>:
<tbody>:
<td>: colspan, headers, rowspan
<tfoot>:

> <th>: abbr, colspan, headers, rowspan, scope

Reviewer 2: `colspan`, `rowspan`, `scope`

<thead>:
<tr>:

<pre>:

> <iframe>: allow, allowfullscreen, allowpaymentrequest, height, name, referrerpolicy, sandbox, src, srcdoc, width

Reviewer 2: - disallow in standard mode.
- `name`, `src`, `width`, `height`, `allowfullscreen`

> # Sanitize HTML

Reviewer 2: What do you think about introducing sanitization "modes" that we can refer to in our yaml schemas?
1. "Loose": the default, applies to markdown files and yaml properties strings with `contentType: markdown`.
2. "Strict": disallows html entirely. We'd start to reference this in [our schemas](https://dev.azure.com/ceapex/Engineering/_git/docs-ui/?path=%2Fpackages%2Fschemas%2Fsrc%2FModule.schema.json&version=GBmatthew%2Fsystemic-focus-colors&line=21&lineEnd=26&lineStartColumn=1&lineEndColumn=5&lineStyle=plain) via `contentType: markdown-strict` or something.

Overall we'd like to discourage use of html in content as it's a constant source of accessibility bugs, layout issues, design regressions.

Reviewer 3: We might also need an special permissive mode for the currently archived content. I don't believe anyone wants to go back and update those pages unless they have to. The permissive mode should not be accessible to other repos though.

Reviewer 2: How about:
1. "Loose": backwards compat... archive repos, certain page layouts like the old hub and landing pages.
2. "Standard": essentially this spec in it's current form
3. "Strict": disallow html

All modes can be activated by metadata/schema contentType.

Author: We can automatically detect hub/landing/archived contents to save us from changing the configs for these repos/files,

[Feature 178608](https://dev.azure.com/ceapex/Engineering/_workitems/edit/178608/)

## Overview

Content authors can embed arbitrary HTML tags to markdown documents. These HTML tags could contain `<script>` tag that allows arbitrary script execution, or `style` attribute that compromises site visual experience.

### Goals

- Remove HTML tags and attributes from user content that may alter site style
- Remove HTML tags and attributes from user content that may allow javascript execution
- Apply HTML sanitization consistently for all user content

### Out of scope

> - Report warnings for removing disallowed HTML tags and attributes

Reviewer 2: Why is this out of scope? Perf? 
1. It would be awesome if these violations were visible to authors.
1. It would be super duper awesome if these violations went into some reporting sink so we could identify and chip away at content using unsupported markup.

From the template/ux perspective we really don't want folks using html at all.

Author: It was in a separate feature. But it make sense to include both in this spec.

- Attribute names starting with `data-`.
- Accessibility attributes: `role` and names starting with `aria-`.
- Standard HTML5 attributes for allowed HTML tag names:
  ```csharp
  "class", "dir", "hidden", "id", "itemid", "itemprop", "itemref", "itemscope", "itemtype,",
  "lang", "part", "slot", "spellcheck", "tabindex", "title", "cite", "value", "reversed",
  "start", "download", "href", "hreflang", "ping", "rel", "target", "type", "datetime",
  "alt", "decoding", "height", "intrinsicsize", "loading", "sizes", "src", "width",
  "abbr", "colspan", "headers", "rowspan", "scope", "allow", "allowfullscreen", "allowpaymentrequest",
  "name", "referrerpolicy", "sandbox", "srcdoc", "role",
  ```

#### HTML5 attribute name allowlist

The HTML5 attribute name allowlist is generated using the following method:

- Based on https://developer.mozilla.org/en-US/docs/Web/HTML/Element.
- Excludes experimental and obsolete attributes.
- Excludes user interactive DOM elements and attributes (like `<button>`, `<menu>`).
- Excludes media DOM elements (like `<video>`, `<audio>`).
- Excludes DOM elements that affect main site structure (like `<body>` and `<header>`)

**Global attributes**

```html

> class, dir, hidden, id,

Reviewer 3: I think some of these attributes should not be in the allowlist. "class", "dir", "hidden", "lang" and "tabindex" all have UI and Accessibility implications that we don't want the authors customizing.

# Sanitize HTML

[Feature 178608](https://dev.azure.com/ceapex/Engineering/_workitems/edit/178608/)

## Overview

Content authors can embed arbitrary HTML tags to markdown documents. These HTML tags could contain `<script>` tag that allows arbitrary script execution, or `style` attribute that compromises site visual experience.

### Goals

- Remove HTML tags and attributes from user content that may alter site style
- Remove HTML tags and attributes from user content that may allow javascript execution

> - Apply HTML sanitization consistently for all user content

Reviewer 4: Does that include all reference contents and YAML contents?

Reviewer 4: If including reference and YAML contents cost much effort, can we consider only including Markdown files as first phase?

Author: The current HTML post processor runs against all user contents, including YAML or reference, this is 0 cost 😄

> ‼️ Due to how the HTML processing pipeline works today, we cannot tell _user HTML_ from _system generated HTML_, so this sanitizer works for both. 
All new `docs-ui` features that depend on build output with custom HTML attributes __MUST__ use `data-` prefix, overwise they'll be stripped by the sanitizer. 

#### Sanitize URL

To prevent script execution like `<a href="javascript://void">`, only `http://` or `https://` are allowed URL schemas for all _external_ links. Absolute URLs (like `/azure`) and relative URLs (like `azure`) are not sanitized.

## Dependencies

The ruleout of this feature impact v3 migration tool since there are likely HTML diffs.
We can either make the change simultaneously in v2 and v3, or change the v3 migration tool to ignore these diffs.

## Appendix

#### HTML5 attribute name allowlist

- This list is based on https://developer.mozilla.org/en-US/docs/Web/HTML/Element.
- This list exclude experimental and obsolete attributes.
- This list exclude user interactive DOM elements and attributes (like `<button>`, `<menu>`).
- This list exclude media DOM elements (like `<video>`, `<audio>`).

> - This list exclude DOM elements that are affect main site structure (like `<body>` and `<header>`)

Reviewer 5: ```suggestion
- This list excludes DOM elements that may affect main site structure (e.g. `<body>` and `<header>`)
```
